### PR TITLE
Post Controls: modularization and reduxification (last attempt, I promise)

### PR DESCRIPTION
**Pull request by Copons**
Follow up from https://github.com/Automattic/wp-calypso/pull/10746
Alternative to https://github.com/Automattic/wp-calypso/pull/10879
Rewrite of https://github.com/Automattic/wp-calypso/pull/10945

- Move the single control items into their own `PostControl` component in order to make `PostControls` more readable.
- Reduxification of user capabilities checks.
- Reduxification of events tracking.

But also:
- Move "show more controls" logic to `PostControls` (as proposed [here](https://github.com/Automattic/wp-calypso/commit/a1cea017c80a33fb3ef177280be86da60b781af0#diff-a3dda1b5d1276d582b092dd2a9e2c818R431)), and therefore revert to a stateful component.
- Keep the Git history of the moved file!

### CODE REVIEW

File: client/my-sites/posts/post-controls/index.jsx [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+				href: `/stats/post/${ post.ID }/${ site.slug }`,
+				icon: 'stats-alt',
+				onClick: recordViewPostStats,
+				text: translate( 'Stats' ),
+			} );
+
+			if ( isEnabled( 'republicize' ) ) {
+				controls.main.push( {
+					className: 'share',
+					disabled: ! this.props.isPublicizeEnabled,
```
Comment (gwwar) [inline client/my-sites/posts/post-controls/index.jsx]: @Copons In a dev env, it looks like the share button is broken. For behavior we can compare with wpcalypso.wordpress.com where republicize is also enabled

**Comment by Copons** [comment]
@gwwar I went with those messages expecting to squash and merge. Do you believe we'd risk losing history again by squashing?

(Anyway, gonna check the sharing, likely I missed something in all that slow committing. 😄 )

**Comment by gwwar** [comment]
I think we can probably fixup to 2 commits, as long as we keep the rename.

**Comment by Copons** [comment]
@gwwar I fiddled around with an awful amount of rebasing, and I ended up with 1 commit for the rename and 1 for everything else. Blaming the new (renamed) file shows the whole history.

WRT to the Share regression: apparently the component wasn't updating on state change.
I manually added [a check](https://github.com/Automattic/wp-calypso/pull/11097/commits/0f33c43756687768e2bd3d8ef9b26bce5a2956a3#diff-a7a83b5a065591f2f1c0ea013feffab1R61) in `shouldComponentUpdate` (which is pretty much the same check used for `showMoreOptions` before moving it in `PostControls`), but this doesn't explain why it is currently working on wpcalypso.

**Comment by gwwar** [comment]
Share button now works, but trash doesn't.

**Comment by gwwar** [comment]
Hmm, still seeing weird behavior with the trash button. Could you verify @Copons?

**Comment by Copons** [comment]
@gwwar You're correct, all actions controlled by the parent component's mixin don't work.
This is what I found:
`PostControls` is half-controlled by post list's `Post` which inherits the [update-post-status mixin](https://github.com/Automattic/wp-calypso/blob/master/client/lib/mixins/update-post-status/index.jsx) that takes care of several actions (trash, delete, restore, publish).
They basically update `Post` internal state on the following properties (among others):
```
{
  previousStatus,
  updatedStatus,
  showMoreOptions
}
```

The third was made obsolete by my attempt of making `PostControl` as self-controlled as possible (and I guess it should be safe to remove it from the mixin).
The other two handle the post status change. For example, when you click on "Trash", `state.updatedStatus` changes into `trashing` (where you see the "trashing" notification) and then into `trash`, which basically hides the post and moves it into the Trashed list.

So: right now (in this branch) the action is performed (e.g. a post gets correctly trashed), but there is no change in the UI.

Now, this is clearly an issue in `Post`'s `shouldComponentUpdate` [method](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/posts/post.jsx#L55-L67).
There you can see very strict checks to prevent excessive renders.

Now for the weirdness.
The only thing related to the state was the `state.showMoreOptions` check ("component should update if next `showMoreOptions` is different from current") - that I removed and later basically replaced with a similar `state.showShare` check.
The mixin sets `showMoreOptions` to false every time the post status changes, so (in `master`) there just cannot be any guarantee that it will update the component.
And yet... the component updates.

If in this branch I try to add
```js
if ( nextState !== this.state ) {
	return true;
}
```
to `shouldComponentUpdate`, all the actions seem to work correctly, but I'm wary of committing this change since I'd rather not kill the post list performance.

**Comment by gwwar** [comment]
Let's cc @retrofox and @aduth if they have any more insight on the shouldComponentUpdate weirdness. I'll try to take some time later and step through.

**Comment by alisterscott** [comment]
Closing as this PR is stale - please reopen if necessary

**Pull request closed without merging**